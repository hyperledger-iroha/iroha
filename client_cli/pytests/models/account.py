"""
This module provides an Account class for working with Iroha network accounts.
"""
from dataclasses import dataclass

@dataclass
class Account:
    """
    Account class represents an Iroha account.

    :param name: The name of the account.
    :type name: str
    :param domain: The domain of the account.
    :type domain: str
    :param public_key: The public key of the account.
    :type public_key: str
    """
    name: str
    domain: str
    public_key: str

    def __repr__(self):
        return f"{self.name}@{self.domain}"
