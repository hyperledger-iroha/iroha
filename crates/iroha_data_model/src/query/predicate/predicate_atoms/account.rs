//! This module contains predicates for account-related objects, mirroring [`crate::account`].

#[cfg(not(feature = "std"))]
use alloc::{format, string::String, vec::Vec};

use iroha_schema::IntoSchema;
use iroha_version::{Decode, Encode};
use serde::{Deserialize, Serialize};

use super::impl_predicate_box;
use crate::{
    account::{Account, AccountId},
    query::{
        predicate::{
            predicate_ast_extensions::AstPredicateExt as _,
            predicate_atoms::{
                domain::DomainIdPredicateBox, MetadataPredicateBox, PublicKeyPredicateBox,
            },
            predicate_combinators::{AndAstPredicate, NotAstPredicate, OrAstPredicate},
            projectors::BaseProjector,
            EvaluatePredicate,
        },
        AstPredicate, CompoundPredicate, HasPredicateBox, HasPrototype,
    },
};

/// A predicate that can be applied to an [`AccountId`].
#[derive(Debug, Clone, PartialEq, Eq, Decode, Encode, Deserialize, Serialize, IntoSchema)]
pub enum AccountIdPredicateBox {
    // object-specific predicates
    /// Checks if the input is equal to the expected value.
    Equals(AccountId),
    // projections
    /// Checks if a predicate applies to the domain ID of the input.
    DomainId(DomainIdPredicateBox),
    /// Checks if a predicate applies to the signatory of the input.
    Signatory(PublicKeyPredicateBox),
}

impl_predicate_box!(AccountId: AccountIdPredicateBox);

impl EvaluatePredicate<AccountId> for AccountIdPredicateBox {
    fn applies(&self, input: &AccountId) -> bool {
        match self {
            AccountIdPredicateBox::Equals(expected) => expected == input,
            AccountIdPredicateBox::DomainId(domain) => domain.applies(&input.domain),
            AccountIdPredicateBox::Signatory(public_key) => public_key.applies(&input.signatory),
        }
    }
}

/// A predicate that can be applied to an [`Account`].
#[derive(Debug, Clone, PartialEq, Eq, Decode, Encode, Deserialize, Serialize, IntoSchema)]
#[allow(unreachable_patterns)]
pub enum AccountPredicateBox {
    // projections
    /// Checks if a predicate applies to the ID of the input.
    Id(AccountIdPredicateBox),
    /// Checks if a predicate applies to the metadata of the input.
    Metadata(MetadataPredicateBox),
}

impl_predicate_box!(Account: AccountPredicateBox);

impl EvaluatePredicate<Account> for AccountPredicateBox {
    fn applies(&self, input: &Account) -> bool {
        match self {
            AccountPredicateBox::Id(id) => id.applies(&input.id),
            AccountPredicateBox::Metadata(metadata) => metadata.applies(&input.metadata),
        }
    }
}

pub mod prelude {
    //! Re-export all predicate boxes for a glob import `(::*)`
    pub use super::{AccountIdPredicateBox, AccountPredicateBox};
}
