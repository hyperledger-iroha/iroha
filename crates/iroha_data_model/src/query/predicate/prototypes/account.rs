//! Account-related prototypes, mirroring types in [`crate::account`].

use super::impl_prototype;
use crate::{
    account::AccountId,
    query::{
        predicate::{
            predicate_atoms::account::{AccountIdPredicateBox, AccountPredicateBox},
            projectors::{
                AccountIdDomainIdProjector, AccountIdProjector, AccountIdSignatoryProjector,
                AccountMetadataProjector, ObjectProjector,
            },
            prototypes::{domain::DomainIdPrototype, MetadataPrototype, PublicKeyPrototype},
        },
        AstPredicate, HasPrototype,
    },
};

/// A prototype of [`AccountId`] for predicate construction.
#[derive(Default, Copy, Clone)]
pub struct AccountIdPrototype<Projector> {
    /// Build a predicate on domain ID of this [`AccountId`]
    pub domain_id: DomainIdPrototype<AccountIdDomainIdProjector<Projector>>,
    /// Build a predicate on signatory of this [`AccountId`]
    pub signatory: PublicKeyPrototype<AccountIdSignatoryProjector<Projector>>,
}

impl_prototype!(AccountIdPrototype: AccountIdPredicateBox);

impl<Projector> AccountIdPrototype<Projector>
where
    Projector: ObjectProjector<Input = AccountIdPredicateBox>,
{
    /// Creates a predicate that checks if the account ID equals the expected value.
    pub fn eq(&self, expected: AccountId) -> Projector::ProjectedPredicate<AccountIdPredicateBox> {
        Projector::project_predicate(AccountIdPredicateBox::Equals(expected))
    }
}

/// A prototype of [`crate::account::Account`] for predicate construction.
#[derive(Default, Copy, Clone)]
pub struct AccountPrototype<Projector> {
    /// Build a predicate on ID of this [`crate::account::Account`]
    pub id: AccountIdPrototype<AccountIdProjector<Projector>>,
    /// Build a predicate on metadata of this [`crate::account::Account`]
    pub metadata: MetadataPrototype<AccountMetadataProjector<Projector>>,
}

impl_prototype!(AccountPrototype: AccountPredicateBox);
