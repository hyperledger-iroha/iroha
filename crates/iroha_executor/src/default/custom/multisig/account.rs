//! Validation and execution logic of instructions for multisig accounts

use super::*;

impl VisitExecute for MultisigRegister {
    fn visit<V: Execute + Visit + ?Sized>(&self, executor: &mut V) {
        let host = executor.host();
        let target_domain = self.account.domain();

        // Any account in a domain can register any multisig account in the domain
        // TODO Restrict access to the multisig signatories?
        // TODO Impose proposal and approval process?
        if target_domain != executor.context().authority.domain() {
            deny!(
                executor,
                "multisig account and its registrant must be in the same domain"
            )
        }

        let Ok(_domain_found) = host
            .query(FindDomains)
            .filter_with(|domain| domain.id.eq(target_domain.clone()))
            .execute_single()
        else {
            deny!(
                executor,
                "domain must exist before registering multisig account"
            );
        };

        for signatory in self.signatories.keys().cloned() {
            let Ok(_signatory_found) = host
                .query(FindAccounts)
                .filter_with(|account| account.id.eq(signatory))
                .execute_single()
            else {
                deny!(
                    executor,
                    "signatories must exist before registering multisig account"
                );
            };
        }
    }

    fn execute<V: Execute + Visit + ?Sized>(self, executor: &mut V) -> Result<(), ValidationFail> {
        let host = executor.host();
        let registrant = executor.context().authority.clone();
        let multisig_account = self.account;
        let multisig_role = multisig_role_for(&multisig_account);

        host.submit(&Register::account(Account::new(multisig_account.clone())))
            .dbg_expect("registrant should successfully register a multisig account");

        host.submit(&SetKeyValue::account(
            multisig_account.clone(),
            SIGNATORIES.parse().unwrap(),
            Json::new(&self.signatories),
        ))
        .dbg_unwrap();

        host.submit(&SetKeyValue::account(
            multisig_account.clone(),
            QUORUM.parse().unwrap(),
            Json::new(&self.quorum),
        ))
        .dbg_unwrap();

        host.submit(&SetKeyValue::account(
            multisig_account.clone(),
            TRANSACTION_TTL_MS.parse().unwrap(),
            Json::new(&self.transaction_ttl_ms),
        ))
        .dbg_unwrap();

        host.submit(&Register::role(
            // Temporarily grant a multisig role to the registrant to delegate the role to the signatories
            Role::new(multisig_role.clone(), registrant.clone()),
        ))
        .dbg_expect("registrant should successfully register a multisig role");

        for signatory in self.signatories.keys().cloned() {
            host.submit(&Grant::account_role(multisig_role.clone(), signatory))
                .dbg_expect(
                    "registrant should successfully grant the multisig role to signatories",
                );
        }

        // FIXME No roles to revoke found, which should have been granted to the registrant
        // host.submit(&Revoke::account_role(multisig_role, registrant))
        //     .dbg_expect(
        //         "registrant should successfully revoke the multisig role from the registrant",
        //     );

        Ok(())
    }
}
