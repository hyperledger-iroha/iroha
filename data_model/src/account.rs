//! Structures, traits and impls related to `Account`s.
#[cfg(not(feature = "std"))]
use alloc::{
    collections::{btree_map, btree_set},
    format,
    string::String,
    vec::Vec,
};
use core::str::FromStr;
#[cfg(feature = "std")]
use std::collections::{btree_map, btree_set};

use derive_more::{DebugCustom, Display};
use getset::Getters;
use iroha_data_model_derive::{model, IdEqOrdHash};
use iroha_primitives::{const_vec::ConstVec, must_use::MustUse};
use iroha_schema::IntoSchema;
use parity_scale_codec::{Decode, Encode};
use serde::{Deserialize, Serialize};
use serde_with::{DeserializeFromStr, SerializeDisplay};

pub use self::model::*;
use crate::{
    asset::{
        prelude::{Asset, AssetId},
        AssetsMap,
    },
    domain::prelude::*,
    metadata::Metadata,
    name::Name,
    HasMetadata, Identifiable, ParseError, PublicKey, Registered,
};

/// API to work with collections of [`Id`] : [`Account`] mappings.
pub type AccountsMap = btree_map::BTreeMap<AccountId, Account>;

// The size of the array must be fixed. If we use more than `1` we
// waste all of that space for all non-multisig accounts. If we
// have 1 signatory per account, we keep the signature on the
// stack. If we have more than 1, we keep everything on the
// heap. Thanks to the union feature, we're not wasting `8Bytes`
// of space, over `Vec`.
type Signatories = btree_set::BTreeSet<PublicKey>;

#[model]
pub mod model {
    use super::*;

    /// Identification of an [`Account`]. Consists of Account name and Domain name.
    ///
    /// # Examples
    ///
    /// ```rust
    /// use iroha_data_model::account::AccountId;
    ///
    /// let id = "user@company".parse::<AccountId>().expect("Valid");
    /// ```
    #[derive(
        DebugCustom,
        Display,
        Clone,
        PartialEq,
        Eq,
        PartialOrd,
        Ord,
        Hash,
        Getters,
        Decode,
        Encode,
        DeserializeFromStr,
        SerializeDisplay,
        IntoSchema,
    )]
    #[display(fmt = "{name}@{domain_id}")]
    #[debug(fmt = "{name}@{domain_id}")]
    #[getset(get = "pub")]
    #[ffi_type]
    pub struct AccountId {
        /// [`Account`]'s [`Domain`](`crate::domain::Domain`) id.
        pub domain_id: DomainId,
        /// [`Account`]'s name.
        pub name: Name,
    }

    /// Account entity is an authority which is used to execute `Iroha Special Instructions`.
    #[derive(
        Debug,
        Display,
        Clone,
        IdEqOrdHash,
        Getters,
        Decode,
        Encode,
        Deserialize,
        Serialize,
        IntoSchema,
    )]
    #[allow(clippy::multiple_inherent_impl)]
    #[display(fmt = "({id})")] // TODO: Add more?
    #[ffi_type]
    pub struct Account {
        /// An Identification of the [`Account`].
        pub id: AccountId,
        /// Assets in this [`Account`].
        pub assets: AssetsMap,
        /// [`Account`]'s signatories.
        pub signatories: Signatories,
        /// Condition which checks if the account has the right signatures.
        #[getset(get = "pub")]
        pub signature_check_condition: SignatureCheckCondition,
        /// Metadata of this account as a key-value store.
        pub metadata: Metadata,
    }

    /// Builder which should be submitted in a transaction to create a new [`Account`]
    #[derive(
        DebugCustom, Display, Clone, IdEqOrdHash, Decode, Encode, Deserialize, Serialize, IntoSchema,
    )]
    #[display(fmt = "[{id}]")]
    #[debug(fmt = "[{id:?}] {{ signatories: {signatories:?}, metadata: {metadata} }}")]
    #[ffi_type]
    pub struct NewAccount {
        /// Identification
        pub id: AccountId,
        /// Signatories, i.e. signatures attached to this message.
        pub signatories: Signatories,
        /// Metadata that should be submitted with the builder
        pub metadata: Metadata,
    }

    /// Condition which checks if the account has the right signatures.
    #[derive(
        Debug,
        Display,
        Clone,
        PartialEq,
        Eq,
        PartialOrd,
        Ord,
        Decode,
        Encode,
        Deserialize,
        Serialize,
        IntoSchema,
    )]
    #[ffi_type(opaque)]
    #[allow(clippy::enum_variant_names)]
    pub enum SignatureCheckCondition {
        #[display(fmt = "AnyAccountSignatureOr({_0:?})")]
        AnyAccountSignatureOr(ConstVec<PublicKey>),
        #[display(fmt = "AllAccountSignaturesAnd({_0:?})")]
        AllAccountSignaturesAnd(ConstVec<PublicKey>),
    }
}

impl AccountId {
    /// Construct [`Self`].
    // NOTE: not derived to preserve order of fields in which [`Self`] is parsed from string
    pub fn new(name: Name, domain_id: DomainId) -> Self {
        Self { domain_id, name }
    }
}

impl Account {
    /// Construct builder for [`Account`] identifiable by [`Id`] containing the given signatories.
    #[inline]
    #[must_use]
    pub fn new(
        id: AccountId,
        signatories: impl IntoIterator<Item = PublicKey>,
    ) -> <Self as Registered>::With {
        <Self as Registered>::With::new(id, signatories)
    }

    /// Get an iterator over [`signatories`](PublicKey) of the `Account`
    #[inline]
    pub fn signatories(&self) -> impl ExactSizeIterator<Item = &PublicKey> {
        self.signatories.iter()
    }

    /// Return a reference to the [`Asset`] corresponding to the asset id.
    #[inline]
    pub fn asset(&self, asset_id: &AssetId) -> Option<&Asset> {
        self.assets.get(asset_id)
    }

    /// Get an iterator over [`Asset`]s of the `Account`
    #[inline]
    pub fn assets(&self) -> impl ExactSizeIterator<Item = &Asset> {
        self.assets.values()
    }

    /// Return `true` if the `Account` contains the given signatory
    #[inline]
    pub fn contains_signatory(&self, signatory: &PublicKey) -> bool {
        self.signatories.contains(signatory)
    }
}

#[cfg(feature = "transparent_api")]
impl Account {
    /// Add [`Asset`] into the [`Account`] returning previous asset stored under the same id
    #[inline]
    pub fn add_asset(&mut self, asset: Asset) -> Option<Asset> {
        self.assets.insert(asset.id().clone(), asset)
    }

    /// Remove asset from the [`Account`] and return it
    #[inline]
    pub fn remove_asset(&mut self, asset_id: &AssetId) -> Option<Asset> {
        self.assets.remove(asset_id)
    }
    /// Add [`signatory`](PublicKey) into the [`Account`].
    ///
    /// If `Account` did not have this signatory present, `true` is returned.
    /// If `Account` did have this signatory present, `false` is returned.
    #[inline]
    pub fn add_signatory(&mut self, signatory: PublicKey) -> bool {
        self.signatories.insert(signatory)
    }

    /// Remove a signatory from the `Account` and return whether the signatory was present in the `Account`
    #[inline]
    pub fn remove_signatory(&mut self, signatory: &PublicKey) -> bool {
        self.signatories.remove(signatory)
    }
}

impl NewAccount {
    fn new(id: AccountId, signatories: impl IntoIterator<Item = PublicKey>) -> Self {
        Self {
            id,
            signatories: signatories.into_iter().collect(),
            metadata: Metadata::default(),
        }
    }

    /// Add [`Metadata`] to the account replacing any previously defined metadata
    #[must_use]
    pub fn with_metadata(mut self, metadata: Metadata) -> Self {
        self.metadata = metadata;
        self
    }
}

impl HasMetadata for NewAccount {
    fn metadata(&self) -> &Metadata {
        &self.metadata
    }
}

impl HasMetadata for Account {
    fn metadata(&self) -> &Metadata {
        &self.metadata
    }
}

impl Registered for Account {
    type With = NewAccount;
}

impl FromIterator<Account> for crate::Value {
    fn from_iter<T: IntoIterator<Item = Account>>(iter: T) -> Self {
        iter.into_iter()
            .map(Into::into)
            .collect::<Vec<Self>>()
            .into()
    }
}

/// Account Identification is represented by `name@domain_name` string.
impl FromStr for AccountId {
    type Err = ParseError;

    fn from_str(string: &str) -> Result<Self, Self::Err> {
        let split = string.rsplit_once('@');
        match split {
            Some(("", _)) => Err(ParseError {
                reason: "`AccountId` cannot be empty",
            }),
            Some((name, domain_id)) if !name.is_empty() && !domain_id.is_empty() => Ok(AccountId {
                name: name.parse()?,
                domain_id: domain_id.parse()?,
            }),
            _ => Err(ParseError {
                reason: "`AccountId` should have format `name@domain_name`",
            }),
        }
    }
}

impl Default for SignatureCheckCondition {
    fn default() -> Self {
        Self::AnyAccountSignatureOr(ConstVec::new_empty())
    }
}

impl SignatureCheckCondition {
    /// Shorthand to create a [`SignatureCheckCondition::AnyAccountSignatureOr`] variant without additional allowed signatures.
    #[inline]
    pub fn any_account_signature() -> Self {
        Self::AnyAccountSignatureOr(ConstVec::new_empty())
    }

    /// Shorthand to create a [`SignatureCheckCondition::AllAccountSignaturesAnd`] variant without additional required signatures.
    #[inline]
    pub fn all_account_signatures() -> Self {
        Self::AllAccountSignaturesAnd(ConstVec::new_empty())
    }

    /// Checks whether the transaction contains all the signatures required by the `SignatureCheckCondition`.
    pub fn check(
        &self,
        account_signatories: &btree_set::BTreeSet<PublicKey>,
        transaction_signatories: &btree_set::BTreeSet<PublicKey>,
    ) -> MustUse<bool> {
        let result = match &self {
            SignatureCheckCondition::AnyAccountSignatureOr(additional_allowed_signatures) => {
                account_signatories
                    .iter()
                    .chain(additional_allowed_signatures.as_ref())
                    .any(|allowed| transaction_signatories.contains(allowed))
            }
            SignatureCheckCondition::AllAccountSignaturesAnd(additional_required_signatures) => {
                account_signatories
                    .iter()
                    .chain(additional_required_signatures.as_ref())
                    .all(|required_signature| transaction_signatories.contains(required_signature))
            }
        };

        MustUse::new(result)
    }
}

/// The prelude re-exports most commonly used traits, structs and macros from this crate.
pub mod prelude {
    pub use super::{Account, AccountId, SignatureCheckCondition};
}

#[cfg(test)]
mod tests {
    use core::cmp::Ordering;

    use iroha_crypto::{KeyPair, PublicKey};

    use super::{AccountId, SignatureCheckCondition};
    use crate::{domain::DomainId, name::Name};

    fn make_key() -> PublicKey {
        KeyPair::generate().unwrap().public_key().clone()
    }

    fn check_signature_check_condition(
        condition: &SignatureCheckCondition,
        account_signatories: &[&PublicKey],
        tx_signatories: &[&PublicKey],
        result: bool,
    ) {
        let account_signatories = account_signatories.iter().copied().cloned().collect();
        let tx_signatories = tx_signatories.iter().copied().cloned().collect();

        assert_eq!(
            condition.check(&account_signatories, &tx_signatories,).0,
            result
        );
    }

    #[test]
    fn signature_check_condition_default() {
        let key1 = make_key();
        let key2 = make_key();
        let key3 = make_key();
        let condition = SignatureCheckCondition::default();

        check_signature_check_condition(&condition, &[], &[], false);
        check_signature_check_condition(&condition, &[&key1], &[], false);
        check_signature_check_condition(&condition, &[], &[&key1], false);
        check_signature_check_condition(&condition, &[&key1], &[&key1], true);
        check_signature_check_condition(&condition, &[&key1], &[&key2], false);
        check_signature_check_condition(&condition, &[&key1, &key2, &key3], &[&key1], true);
        check_signature_check_condition(&condition, &[&key1, &key2, &key3], &[&key2], true);
        check_signature_check_condition(&condition, &[&key1, &key2, &key3], &[&key3], true);
    }

    #[test]
    fn signature_check_condition_all() {
        let key1 = make_key();
        let key2 = make_key();
        let key3 = make_key();
        let condition = SignatureCheckCondition::all_account_signatures();

        // technically, `\forall x \in \emptyset, check(x)` is true for any `check`, so this evaluate to true
        // maybe not the logic we want?
        check_signature_check_condition(&condition, &[], &[], true);
        check_signature_check_condition(&condition, &[], &[&key1], true);

        check_signature_check_condition(&condition, &[&key1], &[], false);
        check_signature_check_condition(&condition, &[&key1], &[&key1], true);
        check_signature_check_condition(&condition, &[&key1], &[&key2], false);
        check_signature_check_condition(&condition, &[&key1, &key2, &key3], &[&key1], false);
        check_signature_check_condition(&condition, &[&key1, &key2, &key3], &[&key2], false);
        check_signature_check_condition(&condition, &[&key1, &key2, &key3], &[&key3], false);
        check_signature_check_condition(&condition, &[&key1, &key2], &[&key1, &key2, &key3], true);
        check_signature_check_condition(&condition, &[&key1, &key2], &[&key1, &key2], true);
        check_signature_check_condition(&condition, &[&key1, &key2], &[&key2, &key3], false);
    }

    #[test]
    fn signature_check_condition_any_or() {
        let key1 = make_key();
        let key2 = make_key();
        let key3 = make_key();
        let condition = SignatureCheckCondition::AnyAccountSignatureOr(vec![key3.clone()].into());

        check_signature_check_condition(&condition, &[], &[], false);
        check_signature_check_condition(&condition, &[], &[&key3], true);
        check_signature_check_condition(&condition, &[], &[&key2], false);
        check_signature_check_condition(&condition, &[], &[&key1, &key2], false);
        check_signature_check_condition(&condition, &[&key2], &[&key2], true);
        check_signature_check_condition(&condition, &[&key2, &key3], &[&key2], true);
        check_signature_check_condition(&condition, &[&key1, &key2], &[&key2], true);
    }

    #[test]
    fn signature_check_condition_all_and() {
        let key1 = make_key();
        let key2 = make_key();
        let key3 = make_key();
        let condition = SignatureCheckCondition::AllAccountSignaturesAnd(vec![key3.clone()].into());

        check_signature_check_condition(&condition, &[], &[], false);
        check_signature_check_condition(&condition, &[], &[&key3], true);
        check_signature_check_condition(&condition, &[&key1], &[&key3], false);
        check_signature_check_condition(&condition, &[&key1], &[&key1, &key3], true);
        check_signature_check_condition(&condition, &[&key2], &[&key1, &key3], false);
        check_signature_check_condition(&condition, &[&key2], &[&key1, &key2, &key3], true);
    }

    #[test]
    fn cmp_account_id() {
        let domain_id_a: DomainId = "a".parse().expect("failed to parse DomainId");
        let domain_id_b: DomainId = "b".parse().expect("failed to parse DomainId");
        let name_a: Name = "a".parse().expect("failed to parse Name");
        let name_b: Name = "b".parse().expect("failed to parse Name");

        let mut account_ids = Vec::new();
        for name in [&name_a, &name_b] {
            for domain_id in [&domain_id_a, &domain_id_b] {
                account_ids.push(AccountId::new(name.clone(), domain_id.clone()));
            }
        }

        for account_id_1 in &account_ids {
            for account_id_2 in &account_ids {
                match (
                    account_id_1.domain_id.cmp(&account_id_2.domain_id),
                    account_id_1.name.cmp(&account_id_2.name),
                ) {
                    // `DomainId` take precedence in comparison
                    // if `DomainId`s are equal than comparison based on `Name`s
                    (Ordering::Equal, ordering) | (ordering, _) => assert_eq!(
                        account_id_1.cmp(account_id_2),
                        ordering,
                        "{account_id_1:?} and {account_id_2:?} are expected to be {ordering:?}"
                    ),
                }
            }
        }
    }
}
