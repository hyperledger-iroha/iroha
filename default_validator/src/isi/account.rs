//! Validation and tokens related to account operations.

use super::*;

tokens!(
    pattern = {
        #[derive(Token, ValidateGrantRevoke, pass_conditions::derive_conversions::account::Owner)]
        #[validate(pass_conditions::account::Owner)]
        pub struct _ {
            pub account_id: <Account as Identifiable>::Id,
        }
    },
    account::tokens: [
        CanUnregisterAccount,
        CanMintUserPublicKeys,
        CanBurnUserPublicKeys,
        CanMintUserSignatureCheckConditions,
        CanSetKeyValueInUserAccount,
        CanRemoveKeyValueInUserAccount,
    ]
);

impl DefaultValidate for Register<Account> {
    fn default_validate<Q>(
        &self,
        _authority: &<Account as Identifiable>::Id,
        _validate_query: Q,
    ) -> Verdict
    where
        Q: Fn(&<Account as Identifiable>::Id, QueryBox) -> Verdict + Copy,
    {
        pass!()
    }
}

impl DefaultValidate for Unregister<Account> {
    fn default_validate<Q>(
        &self,
        authority: &<Account as Identifiable>::Id,
        _validate_query: Q,
    ) -> Verdict
    where
        Q: Fn(&<Account as Identifiable>::Id, QueryBox) -> Verdict + Copy,
    {
        let account_id = self.object_id().clone();

        pass_if!(&account_id == authority);
        pass_if!(tokens::CanUnregisterAccount { account_id }.is_owned_by(authority));

        deny!("Can't unregister another account")
    }
}

impl DefaultValidate for Mint<Account, PublicKey> {
    fn default_validate<Q>(
        &self,
        authority: &<Account as Identifiable>::Id,
        _validate_query: Q,
    ) -> Verdict
    where
        Q: Fn(&<Account as Identifiable>::Id, QueryBox) -> Verdict + Copy,
    {
        let account_id = self.destination_id().clone();

        pass_if!(&account_id == authority);
        pass_if!(tokens::CanMintUserPublicKeys { account_id }.is_owned_by(authority));

        deny!("Can't mint public keys of another account")
    }
}

impl DefaultValidate for Burn<Account, PublicKey> {
    fn default_validate<Q>(
        &self,
        authority: &<Account as Identifiable>::Id,
        _validate_query: Q,
    ) -> Verdict
    where
        Q: Fn(&<Account as Identifiable>::Id, QueryBox) -> Verdict + Copy,
    {
        let account_id = self.destination_id().clone();

        pass_if!(&account_id == authority);
        pass_if!(tokens::CanBurnUserPublicKeys { account_id }.is_owned_by(authority));

        deny!("Can't burn public keys of another account")
    }
}

impl DefaultValidate for Mint<Account, SignatureCheckCondition> {
    fn default_validate<Q>(
        &self,
        authority: &<Account as Identifiable>::Id,
        _validate_query: Q,
    ) -> Verdict
    where
        Q: Fn(&<Account as Identifiable>::Id, QueryBox) -> Verdict + Copy,
    {
        let account_id = self.destination_id().clone();

        pass_if!(&account_id == authority);
        pass_if!(tokens::CanMintUserSignatureCheckConditions { account_id }.is_owned_by(authority));

        deny!("Can't mint signature check conditions of another account")
    }
}

impl DefaultValidate for SetKeyValue<Account, Name, Value> {
    fn default_validate<Q>(
        &self,
        authority: &<Account as Identifiable>::Id,
        _validate_query: Q,
    ) -> Verdict
    where
        Q: Fn(&<Account as Identifiable>::Id, QueryBox) -> Verdict + Copy,
    {
        let account_id = self.object_id().clone();

        pass_if!(&account_id == authority);
        pass_if!(tokens::CanSetKeyValueInUserAccount { account_id }.is_owned_by(authority));

        deny!("Can't set value to the metadata of another account")
    }
}

impl DefaultValidate for RemoveKeyValue<Account, Name> {
    fn default_validate<Q>(
        &self,
        authority: &<Account as Identifiable>::Id,
        _validate_query: Q,
    ) -> Verdict
    where
        Q: Fn(&<Account as Identifiable>::Id, QueryBox) -> Verdict + Copy,
    {
        let account_id = self.object_id().clone();

        pass_if!(&account_id == authority);
        pass_if!(tokens::CanRemoveKeyValueInUserAccount { account_id }.is_owned_by(authority));

        deny!("Can't remove value from the metadata of another account")
    }
}
