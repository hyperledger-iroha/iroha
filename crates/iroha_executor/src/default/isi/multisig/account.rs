//! Validation and execution logic of instructions for multisig accounts

use iroha_executor_data_model::permission::account::CanRegisterAccount;

use super::*;
use crate::permission::domain::is_domain_owner;

impl VisitExecute for MultisigRegister {
    fn visit<V: Execute + Visit + ?Sized>(&self, executor: &mut V) {
        let registrant = executor.context().authority.clone();
        let target_domain = self.account.domain();
        let host = executor.host();

        let Ok(is_domain_owner) = is_domain_owner(target_domain, &registrant, host) else {
            deny!(
                executor,
                "domain must exist before registering multisig account"
            );
        };

        let has_permission = {
            CanRegisterAccount {
                domain: target_domain.clone(),
            }
            .is_owned_by(&registrant, host)
        };

        // Impose the same restriction as for personal account registrations
        // TODO Allow the signatories to register the multisig account? With propose and approve procedures?
        if !(is_domain_owner || has_permission) {
            deny!(
                executor,
                "registrant must have sufficient permission to register an account"
            );
        }

        for signatory in self.signatories.keys().cloned() {
            if host
                .query(FindAccounts)
                .filter_with(|account| account.id.eq(signatory))
                .execute_single()
                .is_err()
            {
                deny!(
                    executor,
                    "signatories must exist before registering multisig account"
                );
            }
        }
    }

    fn execute<V: Execute + Visit + ?Sized>(self, executor: &mut V) -> Result<(), ValidationFail> {
        let domain_owner = executor
            .host()
            .query(FindDomains)
            .filter_with(|domain| domain.id.eq(self.account.domain().clone()))
            .execute_single()
            .dbg_unwrap()
            .owned_by()
            .clone();

        // Authorize as the domain owner:
        // Just having permission to register accounts is insufficient to register multisig roles
        executor.context_mut().authority = domain_owner.clone();

        let multisig_account = self.account;
        let multisig_role = multisig_role_for(&multisig_account);

        visit_seq!(executor
            .visit_register_account(&Register::account(Account::new(multisig_account.clone()))));

        visit_seq!(executor.visit_set_account_key_value(&SetKeyValue::account(
            multisig_account.clone(),
            SIGNATORIES.parse().unwrap(),
            Json::new(&self.signatories),
        )));

        visit_seq!(executor.visit_set_account_key_value(&SetKeyValue::account(
            multisig_account.clone(),
            QUORUM.parse().unwrap(),
            Json::new(self.quorum),
        )));

        visit_seq!(executor.visit_set_account_key_value(&SetKeyValue::account(
            multisig_account.clone(),
            TRANSACTION_TTL_MS.parse().unwrap(),
            Json::new(self.transaction_ttl_ms),
        )));

        visit_seq!(executor.visit_register_role(&Register::role(
            // Temporarily grant a multisig role to the domain owner to delegate the role to the signatories
            Role::new(multisig_role.clone(), domain_owner.clone()),
        )));

        for signatory in self.signatories.keys().cloned() {
            visit_seq!(executor
                .visit_grant_account_role(&Grant::account_role(multisig_role.clone(), signatory)));
        }

        visit_seq!(
            executor.visit_revoke_account_role(&Revoke::account_role(multisig_role, domain_owner))
        );

        Ok(())
    }
}
