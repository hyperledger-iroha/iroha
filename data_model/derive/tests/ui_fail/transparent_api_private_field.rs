use iroha_data_model::account::AccountId;

fn main() {
    let account_id: AccountId = "alice@wonderland".parse().expect("Valid account id");
    println!("ID: {}", account_id.name);
}
