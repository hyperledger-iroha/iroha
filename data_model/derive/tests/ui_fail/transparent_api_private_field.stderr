error[E0616]: field `name` of struct `iroha_data_model::account::AccountId` is private
 --> tests/ui_fail/transparent_api_private_field.rs:5:35
  |
5 |     println!("ID: {}", account_id.name);
  |                                   ^^^^ private field
  |
help: a method `name` also exists, call it with parentheses
  |
5 |     println!("ID: {}", account_id.name());
  |                                       ++
